bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/BufferedHadoopHashTableSource.java

```diff
@@ -0,0 +1,212 @@
+        + "), maxBufferSize="
+        + maxBufferSize;
+  }
+
+  private void writeObject(ObjectOutputStream s) throws IOException {
+    s.writeObject(hashTableSource);
+    s.writeInt(maxBufferSize);
+  }
+
+  private void readObject(ObjectInputStream s) throws IOException, ClassNotFoundException {
+    this.hashTableSource = (HadoopHashTableSource) s.readObject();
+    this.coder = KvCoder.of(StringUtf8Coder.of(), ListCoder.of(RangeHashCoder.of()));
+    this.maxBufferSize = s.readInt();
+  }
+
+  private static class BufferedHashBasedReader extends BoundedReader<KV<String, List<RangeHash>>> {
+
+    private BoundedReader<RangeHash> hashReader;
+    private BufferedHadoopHashTableSource source;
+
+    private List<RangeHash> buffer;
+
+    public BufferedHashBasedReader(
+        BufferedHadoopHashTableSource source, BoundedReader<RangeHash> hashReader) {
+      this.source = source;
+      this.hashReader = hashReader;
+      this.buffer = new ArrayList<>(source.maxBufferSize);
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      if (!hashReader.start()) {
+        // HashReader does not have any hashes, return empty reader.
+        return false;
+      }
+      // Start returned true, consume the current RangeHash.
+      buffer.add(hashReader.getCurrent());
+      bufferRangeHashes();
+      // Buffer is not empty, return true to consume the current buffer.
+      return true;
+    }
+
+    // Reads from hashReader and buffers the RangeHashes.
+    // Returns true if any RangeHashes were read from hashReader.
+    private boolean bufferRangeHashes() throws IOException {
+      boolean readRangeHashes = false;
+      while (buffer.size() < source.maxBufferSize && hashReader.advance()) {
+        readRangeHashes = true;
+        buffer.add(hashReader.getCurrent());
+      }
+      return readRangeHashes;
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      return bufferRangeHashes();
+    }
+
+    @Override
+    public KV<String, List<RangeHash>> getCurrent() {
+      // getCurrent only gets called when buffer is not empty.
+      Preconditions.checkArgument(!buffer.isEmpty(), "Can not get current on empty buffer.");
+      List<RangeHash> hashes = buffer;
+      // Reset the buffer for next batch.
+      buffer = new ArrayList<>(source.maxBufferSize);
+      // GroupBy key is a string and not ImmutableBytesWritable because the WritableCoder is not
+      // deterministic. The outputted PCollection is grouped by the K and needs a deterministic
+      // coder. Having a String K leads to an unfortunate double encoding, ImmutableBytesWritable->
+      // HEX string -> UTF8 encoded string. The number of batches are significantly smaller than
+      // data fetched from Bigtable and should not have meaningful impact on the job performance.
+      return KV.of(Bytes.toStringBinary(hashes.get(0).startInclusive.copyBytes()), hashes);
```

Comment:
consider using `hashes.get(0).startInclusive.hashCode()` for the batch?

Reply:
Yes, the K of KV.

I thought that the key is just used to split the bundles, where does it appear in error messages?
Also, I dont think you really care if there is hash collision, because you are pre-grouping